File: 2018/07-jul/7-29-overloading/swift/Garden.swift
/*
 Struct representing a garden that I do work in at Tod's Point.
 - Author: Andrew Jarombek
 - Date: 7/28/2018
 */

import Foundation

struct Garden {
    
    let animals: [Animal]
    let plants: [Plant]
    let random_objects: [Any]?
    
    /**
     Constructor for a garden with items in it.  These items include animals, plants, and
     optionally other random objects.
     - parameters:
     - withAnimals: An array of animals existing in the garden.
     - plants: An array of plants existing in the garden.
     - andObjects: An array of random objects existing in the garden.  This argument can be ommited (nil)
     */
    init(withAnimals animals: [Animal], plants: [Plant], andObjects objects: [Any]?) {
        self.animals = animals
        self.plants = plants
        random_objects = objects
    }
    
    /**
     Check if an animal exists in the garden.  Internally the animals array
     is checked for existance of the animal
     - parameters:
     - animal: an animal to look for in the garden
     */
    func inGarden(_ animal: Animal) -> Bool {
        return self.animals.contains { $0 == animal }
    }
    
    /**
     Check if a plant exists in the garden.  Internally the plants array
     is checked for existance of the plant
     - parameters:
     - plant: an plant to look for in the garden
     */
    func inGarden(_ plant: Plant) -> Bool {
        return self.plants.contains { $0 == plant }
    }
    
    /**
     Check if a random object exists in the garden.  Internally the random_objects array
     is checked for existance of the object
     - parameters:
     - object: a random object to look for in the garden
     */
    func inGarden(_ object: Any) -> Bool {
        let exists: Bool? = self.random_objects?.contains { $0 as? AnyObject === object as? AnyObject }
        return exists ?? false
    }
}

File: 2018/07-jul/7-29-overloading/csharp/Garden.cs
/**
 * Representation of a Garden containing animals and plants.
 * Author: Andrew Jarombek
 * Date: 1/14/2019
 */

using System;
using System.Collections.Generic;
using System.Linq;

namespace Overload_Override
{
    public class Garden
    {
        // Garden properties
        public List<Animal> Animals { get; }
        public List<Plant> Plants { get; }
        public List<object> RandomObjects { get; }

        /// <summary>
        /// Construct a new garden.  Each garden contains three collections of objects: animals, plants, and
        /// other random objects
        /// </summary>
        /// <param name="animals">The animals living in the garden.  Throws an error if null</param>
        /// <param name="plants">The plants in the garden.  Throws an error if null</param>
        /// <param name="randomObjects">Other random objects in the garden.  Throws an error if null</param>
        /// <exception cref="ArgumentNullException"></exception>
        public Garden(IEnumerable<Animal> animals, IEnumerable<Plant> plants, IEnumerable<object> randomObjects)
        {
            if (animals == null) throw new ArgumentNullException(nameof(animals));
            if (plants == null) throw new ArgumentNullException(nameof(plants));
            if (randomObjects == null) throw new ArgumentNullException(nameof(randomObjects));

            this.Animals = animals.ToList();
            this.Plants = plants.ToList();
            this.RandomObjects = randomObjects.ToList();
        }

        /// <summary>
        /// Determine if an animal lives in the garden
        /// </summary>
        /// <param name="animal">An animal the look in the garden for</param>
        /// <returns>true if the animal exists in the garden, false otherwise</returns>
        public bool InGarden(Animal animal)
        {
            return Animals.Contains(animal);
        }
        
        /// <summary>
        /// Determine if a plant is in the garden
        /// </summary>
        /// <param name="plant">A plant the look in the garden for</param>
        /// <returns>true if the plant exists in the garden, false otherwise</returns>
        public bool InGarden(Plant plant)
        {
            return Plants.Contains(plant);
        }
        
        /// <summary>
        /// Determine if a random object is in the garden
        /// </summary>
        /// <param name="randomObject">A random object to look in the garden for</param>
        /// <returns>true if the object exists in the garden, false otherwise</returns>
        public bool InGarden(object randomObject)
        {
            return RandomObjects.Contains(randomObject);
        }
    }
}